wallet-new/src/Cardano/Wallet/API/V1/Types.hs

```diff
@@ -849,7 +852,50 @@ instance Arbitrary WalletAddress where
                               <*> arbitrary
                               <*> arbitrary
 
-type AccountIndex = Word32
+newtype AccountIndex = AccountIndex { getAccIndex :: Word32 }
+    deriving (Show, Eq, Ord, Generic)
+
+instance Bounded AccountIndex where
+    -- NOTE: minimum for hardened key. See https://iohk.myjetbrains.com/youtrack/issue/CO-309
+    minBound = AccountIndex 2147483648
+    maxBound = AccountIndex maxBound
+
+mkAccountIndex :: Word32 -> Either Text AccountIndex
+mkAccountIndex index | index >= getAccIndex minBound = Right $ AccountIndex index
+                     | otherwise = Left $ sformat
+                            ("mkAccountIndex: Account index should be in range ["%int%".."%int%"]")
+                            (getAccIndex minBound)
+                            (getAccIndex maxBound)
+
+instance ToJSON AccountIndex where
+    toJSON = toJSON . getAccIndex
+
+instance FromJSON AccountIndex where
+    parseJSON =
+        either (fail . toString) pure . mkAccountIndex <=< parseJSON
+
+instance Arbitrary AccountIndex where
+    arbitrary = AccountIndex <$> choose (getAccIndex minBound, getAccIndex maxBound)
+
+deriveSafeBuildable ''AccountIndex
+-- Nothing secret to redact for a AccountIndex.
+instance BuildableSafeGen AccountIndex where
+    buildSafeGen _ = bprint build
```

Comment:
Hmmm... isn't that a bit too recursive ^^" ?

Reply:
uh, you might be right!

I saw a similar definition for `SyncProgress` in the same module https://github.com/input-output-hk/cardano-sl/blob/3acda87dd0730fe0ae66fb302f9ff28a025708c8/wallet-new/src/Cardano/Wallet/API/V1/Types.hs#L694 . Let me have a look at how `BuildableSafeGen` is defined...